capstone/src/test/java/com/google/univiz/scorecard/CollegeDataApiImplTest.java

```diff
@@ -0,0 +1,100 @@
+package com.google.univiz.scorecard;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.assertThrows;
+import static org.mockito.Mockito.when;
+
+import com.google.common.io.Resources;
+import com.google.inject.Guice;
+import com.google.inject.testing.fieldbinder.Bind;
+import com.google.inject.testing.fieldbinder.BoundFieldModule;
+import com.google.univiz.CollegeData;
+import com.google.univiz.GsonModule;
+import com.google.univiz.api.CollegeId;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import javax.inject.Inject;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+@RunWith(JUnit4.class)
+public final class CollegeDataApiImplTest {
+
+  @Rule public final MockitoRule rule = MockitoJUnit.rule();
+  @Bind @Mock private URLProvider mockUrlProvider;
+  @Bind @Mock private CollegeIdReaderProvider mockReaderProvider;
+  private CollegeIdReaderProvider readerProvider = new CollegeIdReaderProviderImpl();
+  @Inject private CollegeDataApiImpl testImpl;
+
+  @Before
+  public void setup() {
+    Guice.createInjector(new GsonModule(), BoundFieldModule.of(this)).injectMembers(this);
+  }
+
+  @Test
+  public void testGetCollegesById() throws IOException {
+    CollegeId collegeId = CollegeId.create(193900);
+    String scorecardUrlString =
+        Resources.getResource(CollegeDataApiImplTest.class, "scorecard_response.json").toString();
+    when(mockUrlProvider.getUrlFromCollegeIds(Arrays.asList(collegeId)))
+        .thenReturn(scorecardUrlString);
+    when(mockReaderProvider.getStreamFromUrl(scorecardUrlString))
+        .then(invocation -> readerProvider.getStreamFromUrl(scorecardUrlString));
```

Comment:
This is a very interesting approach! I expected you to mock the readerProvider and leave url provider as is. Through your use of mocks delegating to an actual instance, you've effectively discovered [spies](https://stackoverflow.com/questions/15052984/what-is-the-difference-between-mocking-and-spying-when-using-mockito).

Here's what I would do to simplify things:

- Delete your mock of `CollegeIdReaderProvider`
- Switch to: `@Bind @Spy private CollegeIdReaderProvider readerProvider = new CollegeIdReaderProviderImpl();`
- Make `CollegeIdReaderProviderImpl` non final s.t. we can spy on it.
- Stop mocking `readerProvider` in all tests except for the IO test.
- Switch the mockito setup in the IOException test to mock out the `readerProvider`.

Reply:
Ah ok this makes it simpler.